evm_arithmetization/src/generation/state.rs

```diff
@@ -77,13 +77,9 @@ pub(crate) trait State<F: Field> {
     fn get_context(&self) -> usize;
 
     /// Checks whether we have reached the maximal cpu length.
-    fn at_end_segment(&self, opt_max_cpu_len: Option<usize>, is_dummy: bool) -> bool {
+    fn at_end_segment(&self, opt_max_cpu_len: Option<usize>) -> bool {
         if let Some(max_cpu_len_log) = opt_max_cpu_len {
-            if is_dummy {
-                self.get_clock() == max_cpu_len_log - NUM_EXTRA_CYCLES_AFTER
-            } else {
-                self.get_clock() == (1 << max_cpu_len_log) - NUM_EXTRA_CYCLES_AFTER
-            }
+            self.get_clock() == (1 << max_cpu_len_log) - NUM_EXTRA_CYCLES_AFTER
```

Comment:
technically outside of this PR, but noticing it only now.
I don't think this is optimized away, and as it is being called in every single iteration of the `run_cpu` internal loop, we'd probably be better off calculating the upper bound `(1 << max_cpu_len_log) - NUM_EXTRA_CYCLES_AFTER` once, and only check `self.get_clock() == cycle_limit` in the loop condition.